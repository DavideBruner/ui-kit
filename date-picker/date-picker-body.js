import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import withMouseOverState from '../hocs/with-mouse-over-state';
import { CalendarIcon, ClockIcon, CloseBoldIcon } from '../icons';
import Spacings from '../materials/spacings';
import styles from './date-picker-body.mod.css';

const getIconTheme = (isDisabled, isMouseOver) => {
  if (isDisabled) return 'grey';
  if (isMouseOver) return 'orange';
  return 'black';
};

export const ClearSection = props => (
  <div
    onClick={props.isDisabled ? undefined : props.onClear}
    className={classnames(styles['clear-icon-container'], {
      [styles['icon-container-disabled']]: props.isDisabled,
      [styles.invalid]: props.isInvalid,
    })}
    onMouseOver={props.handleMouseOver}
    onMouseOut={props.handleMouseOut}
  >
    {!props.isDisabled && (
      <CloseBoldIcon
        size="small"
        theme={getIconTheme(props.isDisabled, props.isMouseOver)}
      />
    )}
  </div>
);
ClearSection.displayName = 'ClearSection';
ClearSection.propTypes = {
  isDisabled: PropTypes.bool,
  isInvalid: PropTypes.bool,
  isMouseOver: PropTypes.bool.isRequired,
  onClear: PropTypes.func,

  // HoC
  handleMouseOut: PropTypes.func.isRequired,
  handleMouseOver: PropTypes.func.isRequired,
};

ClearSection.defaultProps = {
  isInvalid: false,
};

export const ClearSectionWithMouseOverState = withMouseOverState(ClearSection);

export class DatePickerBody extends React.PureComponent {
  static displayName = 'DatePickerBody';

  static propTypes = {
    formattedValue: PropTypes.string,
    isDisabled: PropTypes.bool,
    isInvalid: PropTypes.bool,
    onClearPicker: PropTypes.func,
    placeholder: PropTypes.string,
    size: PropTypes.oneOf(['scale', 'static']),
    timeScale: PropTypes.oneOf(['date', 'time', 'datetime']),
    numberOfFormattedValueChars: PropTypes.number,
  };

  static defaultProps = {
    timeScale: 'date',
    size: 'static',
    isInvalid: false,
    isDisabled: false,
  };

  render() {
    return (
      <Spacings.Inline alignItems="center">
        <div
          className={classnames(styles['date-input-container'], {
            [styles.scale]: this.props.size === 'scale',
          })}
        >
          <input
            className={classnames(styles['date-input'], {
              [styles.invalid]: this.props.isInvalid,
            })}
            placeholder={this.props.placeholder}
            disabled={this.props.isDisabled}
            defaultValue={this.props.formattedValue}
            title={this.props.formattedValue}
            data-input
            size={this.props.numberOfFormattedValueChars}
          />
          <ClearSectionWithMouseOverState
            isDisabled={this.props.isDisabled}
            isInvalid={this.props.isInvalid}
            onClear={this.props.onClearPicker}
          />
          <div
            data-toggle
            className={classnames(styles['calendar-icon-container'], {
              [styles['icon-container-disabled']]: this.props.isDisabled,
              [styles.invalid]: this.props.isInvalid,
            })}
          >
            {this.props.timeScale === 'time' ? (
              <ClockIcon theme={this.props.isDisabled ? 'grey' : 'black'} />
            ) : (
              <CalendarIcon theme={this.props.isDisabled ? 'grey' : 'black'} />
            )}
          </div>
        </div>
      </Spacings.Inline>
    );
  }
}
