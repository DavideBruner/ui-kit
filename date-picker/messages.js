import { defineMessages } from 'react-intl';

export default defineMessages({
  labelRange: {
    id: 'DatePicker.labelRange',
    description: 'Label for range inputs',
    defaultMessage: 'to',
  },
});
