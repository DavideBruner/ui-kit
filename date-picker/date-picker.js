import React from 'react';
import PropTypes from 'prop-types';
import 'flatpickr/dist/themes/airbnb.css';
import Flatpickr from 'flatpickr';
import isTouchDevice from 'is-touch-device';
import { German } from 'flatpickr/dist/l10n/de';
import { compose } from 'recompose';
import { injectIntl } from 'react-intl';
import moment from 'moment-timezone';
import { withUserTimeZone } from '@commercetools-local/application-shell-connectors';
import { parseDateTime } from '@commercetools-local/utils/datetime';
import Constraints from '../materials/constraints';
import { DatePickerBody } from './date-picker-body';
import './date-picker-ct-theme.mod.css';
import styles from './date-picker.mod.css';
import messages from './messages';

const getNumberOfFormattedDateChars = (timeScale, locale, timeZone) => {
  // moment gives us access to its underlying formats for individual locales
  // http://momentjs.com/docs/#/i18n/instance-locale/
  // this allows us to count the number of chars that will be displayed in the
  // formatted date and adjust the input element accordingly
  // Using this technique we can ensure that at least one value is displayed
  switch (timeScale) {
    case 'time':
      return moment()
        .tz(timeZone)
        .locale(locale)
        .localeData()._longDateFormat.LT.length;
    case 'datetime':
      return (
        moment()
          .tz(timeZone)
          .locale(locale)
          .localeData()._longDateFormat.L.length +
        moment()
          .tz(timeZone)
          .locale(locale)
          .localeData()._longDateFormat.LT.length
      );
    case 'date':
      return moment()
        .tz(timeZone)
        .locale(locale)
        .localeData()._longDateFormat.L.length;
    default:
      return 0;
  }
};

export const createFormatter = (timeScale, locale, timeZone) => value => {
  switch (timeScale) {
    case 'time':
      return moment(value, 'HH:mm:ss.SSS')
        .tz(timeZone)
        .locale(locale)
        .format('LT');
    case 'datetime':
      return moment(value)
        .tz(timeZone)
        .locale(locale)
        .format('L LT');
    case 'date':
      return moment(value)
        .tz(timeZone)
        .locale(locale)
        .format('L');
    default:
      return value;
  }
};

export class DatePicker extends React.PureComponent {
  static displayName = 'DatePicker';

  static propTypes = {
    shouldInitializeOnMount: PropTypes.bool,
    isDisabled: PropTypes.bool,
    isInvalid: PropTypes.bool,
    mode: PropTypes.oneOf(['range', 'multiple', 'single']),
    onChange: PropTypes.func.isRequired,
    onClose: PropTypes.func,
    placeholder: PropTypes.string,
    horizontalConstraint: PropTypes.oneOf(['xs', 's', 'm', 'l', 'xl', 'scale']),
    timeScale: PropTypes.oneOf(['date', 'time', 'datetime']),
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.arrayOf(PropTypes.string),
    ]),

    // HoC
    intl: PropTypes.shape({
      formatMessage: PropTypes.func.isRequired,
      locale: PropTypes.string.isRequired,
    }).isRequired,

    // withUserTimeZone
    timeZone: PropTypes.string.isRequired,
  };

  static defaultProps = {
    shouldInitializeOnMount: false,
    isDisabled: false,
    isInvalid: false,
    mode: 'single',
    timeScale: 'date',
    horizontalConstraint: 'scale',
  };

  componentWillMount = () => {
    this.formatter = createFormatter(
      this.props.timeScale,
      this.props.intl.locale,
      this.props.timeZone
    );
    this.numberOfFormattedValueChars = getNumberOfFormattedDateChars(
      this.props.timeScale,
      this.props.intl.locale,
      this.props.timeZone
    );
    this.options = {
      defaultDate: this.props.value,
      enableTime:
        this.props.timeScale === 'time' || this.props.timeScale === 'datetime',
      // flatpickr falls back onto native datetime-inputs on touch-devices
      // these need their values in a standard-format and will format it themselves
      // based on the browsers-locale => using the formatter, will break the fields
      formatDate: isTouchDevice() ? undefined : this.formatter,
      // Gets the corresponding locale. For English we must set it as null.
      // TODO make this asynchronous when more languages available
      locale: this.props.intl.locale.startsWith('de') ? German : null,
      mode: this.props.mode,
      noCalendar: this.props.timeScale === 'time',
      onChange: this.handleChange,
      time_24hr: this.props.intl.locale.startsWith('de'),
      wrap: true,
    };
  };

  // initializing on hove is not feasible for touch-devices, so we init-right away
  // flatpickr does not do its expensive initialization on mobile, so this is safe
  state = {
    initialize: this.props.shouldInitializeOnMount || isTouchDevice(),
  };

  componentDidMount() {
    if (this.shouldInitializeFlatpickr(this.state)) {
      this.initDatepicker();
    }
  }

  shouldInitializeFlatpickr = state => !this.flatpickr && state.initialize;

  componentWillUpdate(nextProps, nextState) {
    if (this.flatpickr && this.props.value !== nextProps.value) {
      this.flatpickr.setDate(nextProps.value, false);
    } else if (this.shouldInitializeFlatpickr(nextState)) {
      this.initDatepicker();
    }
  }

  componentWillUnmount() {
    if (this.flatpickr) {
      this.flatpickr.destroy();
    }
  }

  handleClearPicker = () => {
    if (!this.flatpickr) return;
    this.flatpickr.clear();
    this.flatpickr.jumpToDate();
    this.props.onChange();
  };

  handleChange = selectedDates => {
    switch (this.props.mode) {
      case 'single': {
        const selectedDate =
          selectedDates.length === 0 ? undefined : selectedDates[0];
        this.props.onChange(
          selectedDate &&
            parseDateTime(this.props.timeScale, selectedDate, {
              locale: this.props.intl.locale,
              timeZone: this.props.timeZone,
            })
        );
        break;
      }
      case 'range':
      case 'multiple':
        this.props.onChange(
          selectedDates.map(
            selectedDate =>
              selectedDate &&
              parseDateTime(this.props.timeScale, selectedDate, {
                locale: this.props.intl.locale,
                timeZone: this.props.timeZone,
              })
          )
        );
        break;
      default:
        throw new Error(
          `ui-kit/date-picker: the specified mode '${
            this.props.mode
          }' is not supported.`
        );
    }
  };

  handleMouseOver = () => {
    this.setState(prevState => ({ ...prevState, initialize: true }));
  };

  initDatepicker = () => {
    if (!this.props.isDisabled) {
      const options = {
        onClose: () => {
          if (this.pickerElem) {
            this.pickerElem.blur();
            // NOTE: we need to pass the `value` to enable validations
            // when the picker closes.
            if (this.props.onClose) this.props.onClose(this.props.value);
          }
        },
        ...this.options,
      };

      this.flatpickr = new Flatpickr(this.pickerElem, options);
    }
  };

  getRef = ref => {
    this.pickerElem = ref;
  };

  /**
   * @param  {String} selectedDate This can be `date`, `datetime`, or `time`
   * @return {String} the formatted `selectedDate` based on `timeScale`
   */
  getFormattedValue = selectedDate => {
    // `selectedDate` is expected to be an array
    // when managing `range` and `multiple` modes
    if (this.props.mode === 'multiple')
      return selectedDate.map(v => this.formatter(v)).join(', ');
    if (this.props.mode === 'range')
      return selectedDate
        .map(v => this.formatter(v))
        .join(` ${this.props.intl.formatMessage(messages.labelRange)} `);

    return this.formatter(selectedDate);
  };

  render() {
    return (
      <Constraints.Horizontal constraint={this.props.horizontalConstraint}>
        <div
          className={styles.container}
          onMouseOver={this.handleMouseOver}
          ref={this.getRef}
        >
          <DatePickerBody
            formattedValue={
              this.props.value && this.getFormattedValue(this.props.value)
            }
            isDisabled={this.props.isDisabled}
            isInvalid={this.props.isInvalid}
            onClearPicker={this.handleClearPicker}
            placeholder={this.props.placeholder}
            horizontalConstraint={this.props.horizontalConstraint}
            timeScale={this.props.timeScale}
            numberOfFormattedValueChars={this.numberOfFormattedValueChars}
          />
        </div>
      </Constraints.Horizontal>
    );
  }
}

export default compose(injectIntl, withUserTimeZone())(DatePicker);
