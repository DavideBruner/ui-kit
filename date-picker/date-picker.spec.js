import React from 'react';
import { shallow } from 'enzyme';
import Flatpickr from 'flatpickr';
import isTouchDevice from 'is-touch-device';
import { German } from 'flatpickr/dist/l10n/de';
import moment from 'moment-timezone';
import { DatePicker, createFormatter } from './date-picker';

jest.mock('flatpickr', () => jest.fn());
jest.mock('is-touch-device', () => jest.fn());
jest.mock('moment-timezone');
// let tzMock = jest.fn();
// moment.mockImplementation(() => ({ tz: tzMock }));

const createTestProps = custom => ({
  onChange: jest.fn(),
  value: null,
  placeholder: 'test',
  intl: { formatMessage: jest.fn(message => message.id), locale: 'en' },
  timeZone: 'Europe/Madrid',
  ...custom,
});

describe('<DatePicker />', () => {
  beforeEach(() => {
    isTouchDevice.mockClear();
    isTouchDevice.mockReturnValue(false);
  });
  describe('defaultProps', () => {
    it('should have "single" as `mode`-prop', () => {
      expect(DatePicker.defaultProps.mode).toBe('single');
    });

    it('should have "date" as `timeScale`-prop', () => {
      expect(DatePicker.defaultProps.timeScale).toBe('date');
    });

    it('should have "false" as `isDisabled`-prop', () => {
      expect(DatePicker.defaultProps.isDisabled).toBe(false);
    });

    it('should have "false" as `isInvalid`-prop', () => {
      expect(DatePicker.defaultProps.isInvalid).toBe(false);
    });

    it('should have "scale" as `horizontalConstraint`-prop', () => {
      expect(DatePicker.defaultProps.horizontalConstraint).toBe('scale');
    });
  });

  describe('lifecycle', () => {
    let wrapper;
    let props;

    beforeEach(() => {
      moment.mockImplementation(() => ({
        tz: jest.fn(() => ({
          locale: jest.fn(() => ({
            format: jest.fn(() => '10:00:00.000'),
            localeData: jest.fn(() => ({
              _longDateFormat: {
                LT: ['foo', 'bar'],
                L: ['foo', 'bar'],
              },
            })),
          })),
        })),
      }));
    });

    describe('when shouldInitializeOnMount is true', () => {
      beforeEach(() => {
        props = createTestProps({ shouldInitializeOnMount: true });
        wrapper = shallow(<DatePicker {...props} />);
      });
      it('should have initial state with "initialize" set to true', () => {
        expect(wrapper).toHaveState('initialize', true);
      });
    });

    describe('when shouldInitializeOnMount is false', () => {
      beforeEach(() => {
        props = createTestProps({ shouldInitializeOnMount: false });
        wrapper = shallow(<DatePicker {...props} />);
      });
      it('should have initial state with "initialize" set to false', () => {
        expect(wrapper).toHaveState('initialize', false);
      });
    });

    describe('componentWillMount', () => {
      describe('without locale', () => {
        beforeEach(() => {
          props = createTestProps();
          wrapper = shallow(<DatePicker {...props} />);
          wrapper.instance().componentWillMount();
        });

        it('should create formatter', () => {
          expect(wrapper.instance().formatter).toEqual(expect.any(Function));
        });

        it('should create options object', () => {
          expect(wrapper.instance().options).toEqual({
            defaultDate: null,
            enableTime: false,
            formatDate: expect.any(Function),
            locale: null,
            mode: 'single',
            noCalendar: false,
            onChange: wrapper.instance().handleChange,
            time_24hr: false,
            wrap: true,
          });
        });
      });

      describe('with locale de', () => {
        beforeEach(() => {
          props = createTestProps({
            intl: {
              formatMessage: jest.fn(message => message.id),
              locale: 'de',
            },
          });
          wrapper = shallow(<DatePicker {...props} />);
          wrapper.instance().componentWillMount();
        });

        it('should create options object with the appropriate localization', () => {
          expect(wrapper.instance().options).toEqual(
            expect.objectContaining({ locale: German, time_24hr: true })
          );
        });
      });

      describe('with touch device', () => {
        beforeEach(() => {
          isTouchDevice.mockClear();
          isTouchDevice.mockReturnValue(true);
          props = createTestProps();
          wrapper = shallow(<DatePicker {...props} />);
          wrapper.instance().componentWillMount();
        });

        it('should not set a "formatDate"-function', () => {
          expect(wrapper.instance().options.formatDate).toBe(undefined);
        });
      });

      describe('when is for selecting a "date"', () => {
        beforeEach(() => {
          props = createTestProps({
            timeScale: 'date',
          });
          wrapper = shallow(<DatePicker {...props} />);
          wrapper.instance().componentWillMount();
        });

        it('should not enable the time picker', () => {
          expect(wrapper.instance().options.enableTime).toBe(false);
        });

        it('should display a calendar', () => {
          expect(wrapper.instance().options.noCalendar).toBe(false);
        });
      });

      describe('when is for selecting "date" and "time"', () => {
        beforeEach(() => {
          props = createTestProps({
            timeScale: 'datetime',
          });
          wrapper = shallow(<DatePicker {...props} />);
          wrapper.instance().componentWillMount();
        });

        it('should enable the time picker', () => {
          expect(wrapper.instance().options.enableTime).toBe(true);
        });

        it('should display a calendar', () => {
          expect(wrapper.instance().options.noCalendar).toBe(false);
        });

        describe('when is for selecting "time"', () => {
          beforeEach(() => {
            props = createTestProps({
              timeScale: 'time',
            });
            wrapper = shallow(<DatePicker {...props} />);
            wrapper.instance().componentWillMount();
          });

          it('should enable the time picker', () => {
            expect(wrapper.instance().options.enableTime).toBe(true);
          });

          it('should display a calendar', () => {
            expect(wrapper.instance().options.noCalendar).toBe(true);
          });
        });
      });
    });

    describe('componentDidMount', () => {
      beforeEach(() => {
        Flatpickr.mockClear();
        isTouchDevice.mockClear();
        isTouchDevice.mockReturnValue(true);
        props = createTestProps();
        wrapper = shallow(<DatePicker {...props} />);
        wrapper.instance().componentDidMount();
      });

      it('should initialize flatpickr', () => {
        expect(Flatpickr).toHaveBeenCalledWith(
          undefined,
          expect.objectContaining({
            onClose: expect.any(Function),
          })
        );
      });
    });

    describe('componentWillUpdate', () => {
      describe('when "value"-prop changes on an initialized flatpickr', () => {
        beforeEach(() => {
          props = createTestProps();
          wrapper = shallow(<DatePicker {...props} />);
          wrapper.instance().flatpickr = { setDate: jest.fn() };
          wrapper.instance().componentWillUpdate({ value: '20.10.2018' });
        });

        it('should update the date in flatpickr', () => {
          expect(wrapper.instance().flatpickr.setDate).toHaveBeenCalledWith(
            '20.10.2018',
            false
          );
        });
      });

      describe('when state indicates to initialize', () => {
        describe('when not disabled', () => {
          beforeEach(() => {
            props = createTestProps();
            wrapper = shallow(<DatePicker {...props} />);
            wrapper.instance().componentWillUpdate({}, { initialize: true });
          });

          afterEach(() => {
            Flatpickr.mockClear();
          });

          it('should initialize flatpickr', () => {
            expect(Flatpickr).toHaveBeenCalledWith(undefined, {
              ...wrapper.instance().options,
              onClose: expect.any(Function),
            });
          });
        });

        describe('when disabled', () => {
          beforeEach(() => {
            props = createTestProps({
              isDisabled: true,
            });
            wrapper = shallow(<DatePicker {...props} />);
            wrapper
              .instance()
              .componentWillUpdate({ isDisabled: true }, { initialize: true });
          });

          afterEach(() => {
            Flatpickr.mockClear();
          });

          it('should not initialize flatpickr', () => {
            expect(Flatpickr).not.toHaveBeenCalled();
          });
        });
      });
    });

    describe('componentWillUnmount', () => {
      beforeEach(() => {
        props = createTestProps();
        wrapper = shallow(<DatePicker {...props} />);
        wrapper.instance().flatpickr = { destroy: jest.fn() };
        wrapper.instance().componentWillUnmount();
      });

      it('should destroy the flatpickr-instance', () => {
        expect(wrapper.instance().flatpickr.destroy).toHaveBeenCalled();
      });
    });
  });

  describe('rendering', () => {
    let wrapper;
    let props;

    moment.mockImplementation(() => ({
      tz: jest.fn(() => ({
        locale: jest.fn(() => ({
          format: jest.fn(() => 'some-date'),
          localeData: jest.fn(() => ({
            _longDateFormat: {
              LT: ['foo', 'bar'],
              L: ['foo', 'bar'],
            },
          })),
        })),
      })),
    }));
    beforeEach(() => {
      props = createTestProps();
      wrapper = shallow(<DatePicker {...props} />);
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should contain DatePickerBody', () => {
      expect(wrapper).toRender('DatePickerBody');
    });

    describe('formattedValue', () => {
      let datePickerBody;
      describe('with ranged values', () => {
        beforeEach(() => {
          moment.mockImplementation(date => ({
            tz: jest.fn(() => ({
              locale: jest.fn(() => ({
                format: jest.fn(
                  () => (date.startsWith('2017') ? '11/13/2017' : '11/13/2018')
                ),
                localeData: jest.fn(() => ({
                  _longDateFormat: {
                    LT: ['foo', 'bar'],
                    L: ['foo', 'bar'],
                  },
                })),
              })),
            })),
          }));
          props = createTestProps({
            value: ['2017-11-13T17:36:02.655Z', '2018-11-13T17:36:02.655Z'],
            mode: 'range',
            timeScale: 'date',
          });
          wrapper = shallow(<DatePicker {...props} />);
          datePickerBody = wrapper.find('DatePickerBody');
        });
        it('propagate formattedValue to `DatePickerBody`', () => {
          expect(datePickerBody).toHaveProp(
            'formattedValue',
            '11/13/2017 UIKit.DatePicker.labelRange 11/13/2018'
          );
        });
      });
      describe('with multiple values', () => {
        beforeEach(() => {
          props = createTestProps({
            value: ['2017-11-13T17:36:02.655Z', '2018-11-13T17:36:02.655Z'],
            mode: 'multiple',
            timeScale: 'date',
          });
          wrapper = shallow(<DatePicker {...props} />);
          datePickerBody = wrapper.find('DatePickerBody');
        });
        it('propagate formattedValue to `DatePickerBody`', () => {
          expect(datePickerBody).toHaveProp(
            'formattedValue',
            '11/13/2017, 11/13/2018'
          );
        });
      });
      describe('with single value', () => {
        beforeEach(() => {
          props = createTestProps({
            value: '2017-11-13T17:36:02.655Z',
            mode: 'single',
            timeScale: 'date',
          });
          wrapper = shallow(<DatePicker {...props} />);
          datePickerBody = wrapper.find('DatePickerBody');
        });
        it('propagate formattedValue to `DatePickerBody`', () => {
          expect(datePickerBody).toHaveProp('formattedValue', '11/13/2017');
        });
      });
    });
  });

  describe('callbacks', () => {
    let props;
    let wrapper;
    describe('onClearPicker', () => {
      let flatPickrMock;
      beforeEach(() => {
        flatPickrMock = {
          clear: jest.fn(),
          jumpToDate: jest.fn(),
        };
        props = createTestProps();
        wrapper = shallow(<DatePicker {...props} />);
        wrapper.instance().flatpickr = flatPickrMock;
        wrapper
          .find('DatePickerBody')
          .props()
          .onClearPicker();
      });
      it('should call picker.clear', () => {
        expect(flatPickrMock.clear).toHaveBeenCalledTimes(1);
      });
      it('should call picker.jumpToDate', () => {
        expect(flatPickrMock.jumpToDate).toHaveBeenCalledTimes(1);
      });
      it('should call onChange without a value', () => {
        expect(props.onChange).toHaveBeenCalledTimes(1);
        expect(props.onChange).toHaveBeenLastCalledWith();
      });
    });
  });

  describe('interacting', () => {
    describe('handling change', () => {
      let props;
      let wrapper;

      describe('when in `single` mode', () => {
        beforeEach(() => {
          moment.mockImplementation(() => ({
            tz: jest.fn(() => ({
              locale: jest.fn(() => ({
                format: jest.fn(() => '2017-01-01'),
                localeData: jest.fn(() => ({
                  _longDateFormat: {
                    LT: ['foo', 'bar'],
                    L: ['foo', 'bar'],
                  },
                })),
              })),
            })),
          }));
          props = createTestProps();
          wrapper = shallow(<DatePicker {...props} />);
        });

        describe('with `selectedDates`', () => {
          const selectedDates = [
            new Date('2017-01-01'),
            new Date('2017-02-01'),
          ];
          beforeEach(() => {
            wrapper.instance().handleChange(selectedDates);
          });

          it('should invoke `onChange`', () => {
            expect(props.onChange).toHaveBeenCalled();
          });

          it('should invoke `onChange` with the first date', () => {
            expect(props.onChange).toHaveBeenCalledWith('2017-01-01');
          });
        });

        describe('with empty `selectedDates`', () => {
          beforeEach(() => {
            wrapper.instance().handleChange([]);
          });

          it('should invoke `onChange`', () => {
            expect(props.onChange).toHaveBeenCalled();
          });

          it('should invoke `onChange` with `undefined`', () => {
            expect(props.onChange).toHaveBeenCalledWith(undefined);
          });
        });
      });

      describe('when in `multiple` mode', () => {
        beforeEach(() => {
          moment.mockImplementation(date => ({
            tz: jest.fn(() => ({
              locale: jest.fn(() => ({
                format: jest.fn(() => date),
                localeData: jest.fn(() => ({
                  _longDateFormat: {
                    LT: ['foo', 'bar'],
                    L: ['foo', 'bar'],
                  },
                })),
              })),
            })),
          }));
          props = createTestProps({ mode: 'multiple' });
          wrapper = shallow(<DatePicker {...props} />);
        });

        describe('with filled `selectedDates`', () => {
          const selectedDates = ['2017-01-01', '2017-02-01'];
          beforeEach(() => {
            wrapper.instance().handleChange(selectedDates);
          });

          it('should invoke `onChange`', () => {
            expect(props.onChange).toHaveBeenCalled();
          });

          it('should invoke `onChange` with all dates', () => {
            expect(props.onChange).toHaveBeenCalledWith([
              '2017-01-01',
              '2017-02-01',
            ]);
          });
        });

        describe('with empty `selectedDates`', () => {
          beforeEach(() => {
            wrapper.instance().handleChange([]);
          });

          it('should invoke `onChange`', () => {
            expect(props.onChange).toHaveBeenCalled();
          });

          it('should invoke `onChange` with all dates', () => {
            expect(props.onChange).toHaveBeenCalledWith([]);
          });
        });

        describe('when in `range` mode', () => {
          beforeEach(() => {
            moment.mockImplementation(date => ({
              tz: jest.fn(() => ({
                locale: jest.fn(() => ({
                  format: jest.fn(() => date),
                  localeData: jest.fn(() => ({
                    _longDateFormat: {
                      LT: ['foo', 'bar'],
                      L: ['foo', 'bar'],
                    },
                  })),
                })),
              })),
            }));
            props = createTestProps({ mode: 'range' });
            wrapper = shallow(<DatePicker {...props} />);
          });

          describe('with filled `selectedDates`', () => {
            const selectedDates = ['2017-01-01', '2017-02-01'];
            beforeEach(() => {
              wrapper.instance().handleChange(selectedDates);
            });

            it('should invoke `onChange`', () => {
              expect(props.onChange).toHaveBeenCalled();
            });

            it('should invoke `onChange` with all dates', () => {
              expect(props.onChange).toHaveBeenCalledWith([
                '2017-01-01',
                '2017-02-01',
              ]);
            });
          });

          describe('with empty `selectedDates`', () => {
            beforeEach(() => {
              wrapper.instance().handleChange([]);
            });

            it('should invoke `onChange`', () => {
              expect(props.onChange).toHaveBeenCalled();
            });

            it('should invoke `onChange` with all dates', () => {
              expect(props.onChange).toHaveBeenCalledWith([]);
            });
          });
        });
      });
    });

    describe('handling hover', () => {
      let props;
      let wrapper;

      describe('when hovering over wrapper', () => {
        beforeEach(() => {
          props = createTestProps();
          wrapper = shallow(<DatePicker {...props} />);
          wrapper.instance().handleMouseOver();
        });

        it('should change state for initialize to true', () => {
          expect(wrapper.state('initialize')).toBe(true);
        });
      });
    });
  });
});

describe('createFormatter', () => {
  let formatter;
  const timeZone = 'Europe/Madrid';

  describe('with locale de', () => {
    describe('with date', () => {
      beforeEach(() => {
        moment.mockImplementation(() => ({
          tz: jest.fn(() => ({
            locale: jest.fn(() => ({
              format: jest.fn(() => '31.08.2017'),
              localeData: jest.fn(() => ({
                _longDateFormat: {
                  LT: ['foo', 'bar'],
                  L: ['foo', 'bar'],
                },
              })),
            })),
          })),
        }));
        formatter = createFormatter('date', 'de', timeZone);
      });

      it('should format the date', () => {
        expect(formatter('2017-08-31T23:22:12.811')).toEqual('31.08.2017');
      });
    });

    describe('with datetime', () => {
      beforeEach(() => {
        moment.mockImplementation(() => ({
          tz: jest.fn(() => ({
            locale: jest.fn(() => ({
              format: jest.fn(() => '31.08.2017 23:22'),
              localeData: jest.fn(() => ({
                _longDateFormat: {
                  LT: ['foo', 'bar'],
                  L: ['foo', 'bar'],
                },
              })),
            })),
          })),
        }));
        formatter = createFormatter('datetime', 'de', timeZone);
      });

      it('should format the datetime', () => {
        expect(formatter('2017-08-31T23:22:12.811')).toEqual(
          '31.08.2017 23:22'
        );
      });
    });

    describe('with time', () => {
      beforeEach(() => {
        moment.mockImplementation(() => ({
          tz: jest.fn(() => ({
            locale: jest.fn(() => ({
              format: jest.fn(() => '23:22'),
              localeData: jest.fn(() => ({
                _longDateFormat: {
                  LT: ['foo', 'bar'],
                  L: ['foo', 'bar'],
                },
              })),
            })),
          })),
        }));
        formatter = createFormatter('time', 'de', timeZone);
      });

      it('should format the datetime', () => {
        expect(formatter('23:22:12.811')).toEqual('23:22');
      });
    });
  });

  describe('with locale en', () => {
    describe('with date', () => {
      beforeEach(() => {
        moment.mockImplementation(() => ({
          tz: jest.fn(() => ({
            locale: jest.fn(() => ({
              format: jest.fn(() => '08/31/2017'),
              localeData: jest.fn(() => ({
                _longDateFormat: {
                  LT: ['foo', 'bar'],
                  L: ['foo', 'bar'],
                },
              })),
            })),
          })),
        }));
        formatter = createFormatter('date', 'en', timeZone);
      });

      it('should format the date', () => {
        expect(formatter('2017-08-31T23:22:12.811')).toEqual('08/31/2017');
      });
    });

    describe('with datetime', () => {
      beforeEach(() => {
        moment.mockImplementation(() => ({
          tz: jest.fn(() => ({
            locale: jest.fn(() => ({
              format: jest.fn(() => '08/31/2017 11:22 PM'),
              localeData: jest.fn(() => ({
                _longDateFormat: {
                  LT: ['foo', 'bar'],
                  L: ['foo', 'bar'],
                },
              })),
            })),
          })),
        }));
        formatter = createFormatter('datetime', 'en', timeZone);
      });

      it('should format the datetime', () => {
        expect(formatter('2017-08-31T23:22:12.811')).toEqual(
          '08/31/2017 11:22 PM'
        );
      });
    });

    describe('with time', () => {
      beforeEach(() => {
        moment.mockImplementation(() => ({
          tz: jest.fn(() => ({
            locale: jest.fn(() => ({
              format: jest.fn(() => '11:22 PM'),
              localeData: jest.fn(() => ({
                _longDateFormat: {
                  LT: ['foo', 'bar'],
                  L: ['foo', 'bar'],
                },
              })),
            })),
          })),
        }));
        formatter = createFormatter('time', 'en', timeZone);
      });

      it('should format the datetime', () => {
        expect(formatter('23:22:12.811')).toEqual('11:22 PM');
      });
    });
  });
});

describe('moment integration', () => {
  describe('when accessing localeData', () => {
    beforeEach(() => {
      moment.mockImplementation(() => ({
        tz: jest.fn(() => ({
          locale: jest.fn(() => ({
            format: jest.fn(() => '08/31/2017 11:22 PM'),
            localeData: jest.fn(() => ({
              _longDateFormat: {
                LT: ['foo', 'bar'],
                L: ['foo', 'bar'],
              },
            })),
          })),
        })),
      }));
    });
    it('should have a format for LT', () => {
      expect(
        moment()
          .tz('Europe/Madrid')
          .locale('de')
          .localeData()._longDateFormat.LT
      ).toBeDefined();
    });
    it('should have a format for L', () => {
      expect(
        moment()
          .tz('Europe/Madrid')
          .locale('de')
          .localeData()._longDateFormat.L
      ).toBeDefined();
    });
  });
});
