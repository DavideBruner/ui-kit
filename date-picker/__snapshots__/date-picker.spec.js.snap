// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatePicker /> rendering should match snapshot 1`] = `
<div
  className=""
  onMouseOver={[Function]}
>
  <DatePickerBody
    formattedValue={null}
    isDisabled={false}
    isInvalid={false}
    numberOfFormattedValueChars={10}
    onClearPicker={[Function]}
    placeholder="test"
    size="static"
    timeScale="date"
  />
</div>
`;
